electroncash/rpa/paycode.py

```diff
@@ -243,15 +243,16 @@ def generate_transaction_from_paycode(wallet, config, amount, rpa_paycode=None,
         # Generate the raw transaction
         raw_tx_string = tx.as_dict()["hex"]
 
-        # Get the TxId for this raw Tx.
-        double_hash_tx = bytearray(sha256(sha256(bytes.fromhex(raw_tx_string))))
-        double_hash_tx.reverse()
-        txid = double_hash_tx.hex()
+        # Get the 'txinid' (double hash of serialized input) for this raw Tx.
+        txin= tx._inputs[0]
+        txin_script = Transaction.input_script(txin,False,sign_schnorr = True)
+        serialized_signed_input = Transaction.serialize_input(txin,txin_script)
+        double_hash_txin = bytearray(sha256(sha256(bytes.fromhex(serialized_signed_input))))
+        txinid = double_hash_tx.hex()
 
         # Check if we got a successful match.  If so, exit.
-        if txid[0:prefix_chars].upper() == paycode_field_scan_pubkey[2:prefix_chars + 2].upper():
-            print_msg("Grinding successful after ", grind_nonce, " iterations.")
-            print_msg("Transaction Id: ", txid)
+        if txinid[0:prefix_chars].upper() == paycode_field_scan_pubkey[2:prefix_chars + 2].upper():
```

Comment:
So `prefix_chars` refers to 4-bit chunks at a time? Or?  I guess this is correct then...?

Reply:
You should remove the `.upper()` calls for both items.  They waste cycles needlessly.  Python hex is always lowercase, always.  You can take that to the bank.

Reply:
Also this is wrong:     `elif paycode_field_prefix_size == "0C":` up here: https://github.com/Electron-Cash/Electron-Cash/pull/2029/files#diff-3476eab339a7a5fc0b43507b5646ecd00e5c986326180fb5ada1b1d4f4b00397L191

Reply:
The paycode itself may have upper case hex values, so we should convert either to lower or upper case.  if upper case bothers you can i can make it lower case but it shouldnt be a big deal i think?

Reply:
Ok:

```python
    # Decode the paycode
    rprefix, addr_hash = addr.decode(rpa_paycode)
    paycode_hex = addr_hash.hex()  # <--- this is generated by python and is guaranteed to be lowercase
```

then a few lines later:

```python
    elif paycode_field_prefix_size == "0C":  # <--- this will NEVER be "0C".. it will always be "0c"
```

Reply:
Oh wait I misread your code.. you are doing this:

```python
    paycode_field_prefix_size = paycode_hex[2:4].upper()`
```

Ok this is correct at least!!  Scratch the claim it won't work, it will..


I mean.. you can do that there, and then also do `.upper()` in the line for this conversation -- it's just wasted cycles.  I agree that grinding needs optimization but you can ... *not* call `.upper()` and rely on the fact that the data is already `.lower()` with certainty.

All the data being examined has already gone through `.hex()` which returns `lower`..  that is a guarantee..  so why then do `.upper()`?

Reply:
> The paycode itself may have upper case hex values, s

Yeah but this code never looks at the input as it came it from use -- it is using internal values having been processed by `.hex()`....